### PR TITLE
Issue 759: This will add the cleanup service

### PR DESCRIPTION
**Pull request by bailaC**

### CODE REVIEW

**Comment by bailaC** [comment]
@destogl Please check the draft PR. I think the main logic of the transition is missing. I will check again at my end.

**Comment by saikishor** [comment]
Hello @bailaC,

After looking roughly at the `cleanup_controller` service, it seems to me similar to the `unload_controller` service that exist in the controller manager.

https://github.com/ros-controls/ros2_control/blob/89c8658a3419e4d23aea37eda7b17a0d1eecb840/controller_manager/src/controller_manager.cpp#L601-L661

What do you think?

**Review by saikishor — COMMENTED** [review]
Hello @bailaC,

After looking roughly at the `cleanup_controller` service, it seems to me similar to the `unload_controller` service that exist in the controller manager.

https://github.com/ros-controls/ros2_control/blob/89c8658a3419e4d23aea37eda7b17a0d1eecb840/controller_manager/src/controller_manager.cpp#L601-L661

What do you think?

**Review by destogl — COMMENTED** [review]
@bailaC can you please write tests for the newly added code.

**Review by saikishor — CHANGES_REQUESTED** [review]
Apart from the following changes, I want to know your opinion @destogl @bailaC on the following

Does it make sense to call the service alone `unconfigure_controller` instead of `cleanup_controller` (We can leave the rest of the methods without renaming as internally it's fine)?. Because it might give more or less the basic idea of the service in the context of controllers.

**Comment by soham2560** [comment]
suggested method to commit to this pr? can I directly push to [StoglRobotics-forks:issue-759](https://github.com/StoglRobotics-forks/ros2_control/tree/issue-759) or create a PR targetting it instead

**Comment by christophfroehlich** [comment]
> @christophfroehlich suggested method to commit to this pr? should I create a PR targetting StoglRobotics-forks:issue-759](https://github.com/StoglRobotics-forks/ros2_control/tree/issue-759) or something else

First, get familiar with the code and discuss the open points of this PR above. Then, create a fork of StoglRobotics-forks:issue-759, but open a PR against ros-controls:master. So the original authors of the earlier commits will remain, and we'll just close this PR here.

**Comment by soham2560** [comment]
> Apart from the following changes, I want to know your opinion @destogl @bailaC on the following
 Does it make sense to call the service alone unconfigure_controller instead of cleanup_controller (We can leave the rest of the methods without renaming as internally it's fine)?. Because it might give more or less the basic idea of the service in the context of controllers.
 
@saikishor  considering the conversation at https://github.com/ros-controls/ros2_control/pull/2163 would you say `cleanup_controller` as a name for the service would be fine now? since the state it goes to is clarified by the paranthesis addition in gui

File: controller_manager/src/controller_manager.cpp [hunk 1]
```diff
@@ -619,52 +635,109 @@ controller_interface::ControllerInterfaceBaseSharedPtr ControllerManager::load_c
...
+  if (found_it == to.end())
+  {
+    // Fails if we could not remove the controllers
+    to.clear();
+    RCLCPP_ERROR(
+      get_logger(),
+      "Could not unload controller with name '%s' because no controller with this name exists",
+      controller_name.c_str());
+    return controller_interface::return_type::ERROR;
+  }
```
[the remark this answers is not in the document]
Comment (soham2560) [reply controller_manager/src/controller_manager.cpp:712]: makes sense actually, I will change it once we clear up the prior conversations

**Comment by soham2560** [comment]
Have added draft PR at https://github.com/ros-controls/ros2_control/pull/2414 as suggested in https://github.com/ros-controls/ros2_control/pull/1236#issuecomment-2813793729, some changes remaining but is ready to for a look.

@saikishor I will be making the `unconfigure_controller` changes we talked about, however what about the other suggestions you have put up, I agree with most of them, should I make the changes directly in #2414 or would you like to do a full review there directly?

**Comment by christophfroehlich** [comment]
Closing in favor of #2414

**Comment by saikishor** [comment]
> Have added draft PR at #2414 as suggested in [#1236 (comment)](https://github.com/ros-controls/ros2_control/pull/1236#issuecomment-2813793729), some changes remaining but is ready to for a look.
> 
> @saikishor I will be making the `unconfigure_controller` changes we talked about, however what about the other suggestions you have put up, I agree with most of them, should I make the changes directly in #2414 or would you like to do a full review there directly?

Yes, please apply them :)

**Pull request closed without merging**